File: java-service/src/main/java/com/testmind/controller/UserController.java
package com.testmind.controller;

import com.testmind.dto.UserDTO;
import com.testmind.dto.CreateUserRequest;
import com.testmind.dto.UpdateUserRequest;
import com.testmind.dto.ApiResponse;
import com.testmind.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import jakarta.validation.Valid;
import java.util.List;

@RestController
@RequestMapping("/api/v1/users")
@CrossOrigin(origins = "*")
public class UserController {

    private final UserService userService;

    @Autowired
    public UserController(UserService userService) {
        this.userService = userService;
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<UserDTO>>> getAllUsers() {
        List<UserDTO> users = userService.findAllUsers();
        return ResponseEntity.ok(ApiResponse.success(users));
    }

    @GetMapping("/{id}")
    public ResponseEntity<ApiResponse<UserDTO>> getUserById(@PathVariable Long id) {
        UserDTO user = userService.findUserById(id);
        return ResponseEntity.ok(ApiResponse.success(user));
    }

    @GetMapping("/email/{email}")
    public ResponseEntity<ApiResponse<UserDTO>> getUserByEmail(@PathVariable String email) {
        UserDTO user = userService.findUserByEmail(email);
        return ResponseEntity.ok(ApiResponse.success(user));
    }

    @GetMapping("/username/{username}")
    public ResponseEntity<ApiResponse<UserDTO>> getUserByUsername(@PathVariable String username) {
        UserDTO user = userService.findUserByUsername(username);
        return ResponseEntity.ok(ApiResponse.success(user));
    }

    @GetMapping("/search")
    public ResponseEntity<ApiResponse<List<UserDTO>>> searchUsers(@RequestParam String keyword) {
        List<UserDTO> users = userService.searchUsers(keyword);
        return ResponseEntity.ok(ApiResponse.success(users));
    }

    @GetMapping("/active")
    public ResponseEntity<ApiResponse<List<UserDTO>>> getActiveUsers() {
        List<UserDTO> users = userService.findActiveUsers();
        return ResponseEntity.ok(ApiResponse.success(users));
    }

    @GetMapping("/active/count")
    public ResponseEntity<ApiResponse<Long>> getActiveUserCount() {
        Long count = userService.getActiveUserCount();
        return ResponseEntity.ok(ApiResponse.success(count));
    }

    @GetMapping("/recent")
    public ResponseEntity<ApiResponse<List<UserDTO>>> getRecentUsers() {
        List<UserDTO> users = userService.findRecentUsers();
        return ResponseEntity.ok(ApiResponse.success(users));
    }

    @PostMapping
    public ResponseEntity<ApiResponse<UserDTO>> createUser(@Valid @RequestBody CreateUserRequest request) {
        UserDTO user = userService.createUser(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.success(user));
    }

    @PutMapping("/{id}")
    public ResponseEntity<ApiResponse<UserDTO>> updateUser(@PathVariable Long id, @Valid @RequestBody UpdateUserRequest request) {
        UserDTO user = userService.updateUser(id, request);
        return ResponseEntity.ok(ApiResponse.success(user));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<ApiResponse<Void>> deleteUser(@PathVariable Long id) {
        userService.deleteUser(id);
        return ResponseEntity.ok(ApiResponse.success(null));
    }

    @PatchMapping("/{id}/activate")
    public ResponseEntity<ApiResponse<UserDTO>> activateUser(@PathVariable Long id) {
        UserDTO user = userService.activateUser(id);
        return ResponseEntity.ok(ApiResponse.success(user));
    }

    @PatchMapping("/{id}/deactivate")
    public ResponseEntity<ApiResponse<UserDTO>> deactivateUser(@PathVariable Long id) {
        UserDTO user = userService.deactivateUser(id);
        return ResponseEntity.ok(ApiResponse.success(user));
    }
}

File: node-service/src/controllers/userController.js
/**
 * User controller for TestMind API
 */

const userService = require('../services/userService');
const { User, CreateUserRequest, UpdateUserRequest, ApiResponse } = require('../models');

exports.getAllUsers = async (req, res) => {
  try {
    const users = await userService.findAllUsers();
    res.json(ApiResponse.success(users));
  } catch (error) {
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.getUserById = async (req, res) => {
  try {
    const { id } = req.params;
    const user = await userService.findUserById(parseInt(id));
    if (!user) {
      return res.status(404).json(ApiResponse.error('User not found'));
    }
    res.json(ApiResponse.success(user));
  } catch (error) {
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.getUserByEmail = async (req, res) => {
  try {
    const { email } = req.params;
    const user = await userService.findUserByEmail(email);
    if (!user) {
      return res.status(404).json(ApiResponse.error('User not found'));
    }
    res.json(ApiResponse.success(user));
  } catch (error) {
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.getUserByUsername = async (req, res) => {
  try {
    const { username } = req.params;
    const user = await userService.findUserByUsername(username);
    if (!user) {
      return res.status(404).json(ApiResponse.error('User not found'));
    }
    res.json(ApiResponse.success(user));
  } catch (error) {
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.searchUsers = async (req, res) => {
  try {
    const { keyword } = req.query;
    const users = await userService.searchUsers(keyword);
    res.json(ApiResponse.success(users));
  } catch (error) {
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.getActiveUsers = async (req, res) => {
  try {
    const users = await userService.findActiveUsers();
    res.json(ApiResponse.success(users));
  } catch (error) {
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.getRecentUsers = async (req, res) => {
  try {
    const users = await userService.findRecentUsers();
    res.json(ApiResponse.success(users));
  } catch (error) {
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.getUserCount = async (req, res) => {
  try {
    const count = await userService.getUserCount();
    res.json(ApiResponse.success(count));
  } catch (error) {
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.createUser = async (req, res) => {
  try {
    const userData = req.body;
    const user = await userService.createUser(userData);
    res.status(201).json(ApiResponse.success(user));
  } catch (error) {
    if (error.message.includes('already exists')) {
      return res.status(409).json(ApiResponse.error(error.message));
    }
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.updateUser = async (req, res) => {
  try {
    const { id } = req.params;
    const userData = req.body;
    const user = await userService.updateUser(parseInt(id), userData);
    res.json(ApiResponse.success(user));
  } catch (error) {
    if (error.message.includes('not found')) {
      return res.status(404).json(ApiResponse.error(error.message));
    }
    if (error.message.includes('already exists')) {
      return res.status(409).json(ApiResponse.error(error.message));
    }
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.deleteUser = async (req, res) => {
  try {
    const { id } = req.params;
    await userService.deleteUser(parseInt(id));
    res.json(ApiResponse.success(null));
  } catch (error) {
    if (error.message.includes('not found')) {
      return res.status(404).json(ApiResponse.error(error.message));
    }
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.activateUser = async (req, res) => {
  try {
    const { id } = req.params;
    const user = await userService.activateUser(parseInt(id));
    res.json(ApiResponse.success(user));
  } catch (error) {
    if (error.message.includes('not found')) {
      return res.status(404).json(ApiResponse.error(error.message));
    }
    res.status(500).json(ApiResponse.error(error.message));
  }
};

exports.deactivateUser = async (req, res) => {
  try {
    const { id } = req.params;
    const user = await userService.deactivateUser(parseInt(id));
    res.json(ApiResponse.success(user));
  } catch (error) {
    if (error.message.includes('not found')) {
      return res.status(404).json(ApiResponse.error(error.message));
    }
    res.status(500).json(ApiResponse.error(error.message));
  }
};
